src/storage/src/source/kafka.rs

```diff
@@ -215,18 +215,32 @@ impl SourceConnectionBuilder for KafkaSourceConnection {
                 // Default value obtained from https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
                 .unwrap_or_else(|| Duration::from_secs(300));
 
+            let status_report = Arc::clone(&health_status);
+
             thread::Builder::new()
                 .name("kafka-metadata".to_string())
                 .spawn(move || {
+                    info!(
+                        refresh_frequency =? metadata_refresh_frequency,
+                        "starting kafka metadata refresh thread"
+                    );
                     while let Some(partition_info) = partition_info.upgrade() {
-                        match get_kafka_partitions(&consumer, &topic, Duration::from_secs(30)) {
+                        let result =
+                            get_kafka_partitions(&consumer, &topic, Duration::from_secs(30));
+                        match result {
                             Ok(info) => {
                                 *partition_info.lock().unwrap() = Some(info);
+                                *status_report.lock().unwrap() = Some(HealthStatus::Running);
                                 thread::park_timeout(metadata_refresh_frequency);
                             }
-                            Err(_) => thread::park_timeout(Duration::from_secs(30)),
+                            Err(e) => {
```

Comment:
This replaces reporting via the `GlueConsumerContext`, right? Are we sure that the same messages that would shop up there (in the glue context) show up here?

Reply:
> This replaces reporting via the `GlueConsumerContext`, right?

Yes!

> Are we sure that the same messages that would shop up there (in the glue context) show up here?

No! The most common and significant errors (network issues, auth, broker unreachable, misconfigured topic, etc.) can be spotted both ways, but there's no guarantee that they see exactly the same issues. Which is not so say that this will detect fewer errors! I think it's probably a wash in that respect.

However this implementation is a lot better at detecting _recovery_, which is just as important.

Reply:
could we just do both? still report errors from the glue context, but rely on this loop to detect recovery?

((Sorry if this is stalling the PR, I think it's great to merge as is, and we can discuss after that.))

Reply:
Cool! Let's talk it through in our meeting later.

Reply:
I did some further testing on ACL errors:
- Denying `describe` operations is both logged and captured by the metadata poll.
- Denying `read` operations is neither. This is surprising to me! But at least the new implementation is no worse than the old.

I may cut a ticket to track read permissions errors, but I think that's all the followup I have for now otherwise.

Reply:
👍